src/lib/server/postgres_protocol_handler.cpp

```diff
@@ -0,0 +1,266 @@
+      _write_buffer.template put_value<uint32_t>(static_cast<uint32_t>(value_string.value().size()));
+
+      // Text mode means all values are sent as non-terminated strings
+      _write_buffer.put_string(value_string.value(), IgnoreNullTerminator::Yes);
+    } else {
+      // NULL values are represented by setting the value's length to -1
+      _write_buffer.template put_value<int32_t>(-1);
+    }
+  }
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::send_command_complete(const std::string& command_complete_message) {
+  const auto packet_size = sizeof(LENGTH_FIELD_SIZE) + command_complete_message.size() + 1u /* null terminator */;
+  _write_buffer.template put_value(PostgresMessageType::CommandComplete);
+  _write_buffer.template put_value<uint32_t>(static_cast<uint32_t>(packet_size));
+  _write_buffer.put_string(command_complete_message);
+}
+
+template <typename SocketType>
+std::pair<std::string, std::string> PostgresProtocolHandler<SocketType>::read_parse_packet() {
+  _read_buffer.template get_value<uint32_t>();  // Ignore packet size
+
+  const std::string statement_name = _read_buffer.get_string();
+  const std::string query = _read_buffer.get_string();
+
+  // The number of parameter data types specified (can be zero). These data types are ignored currently.
+  const auto data_types_specified = _read_buffer.template get_value<uint16_t>();
+
+  for (auto i = 0; i < data_types_specified; i++) {
+    // Specifies the object ID of the parameter data type.
+    // Placing a zero here is equivalent to leaving the type unspecified.
+    /*auto data_type = */ _read_buffer.template get_value<int32_t>();
+  }
+
+  return {statement_name, query};
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::read_sync_packet() {
+  // This packet has no body. Hence, only read and ignore its size.
+  _read_buffer.template get_value<uint32_t>();
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::send_status_message(const PostgresMessageType message_type) {
+  _write_buffer.template put_value(message_type);
+  _write_buffer.template put_value<uint32_t>(LENGTH_FIELD_SIZE);
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::read_describe_packet() {
+  const auto packet_length = _read_buffer.template get_value<uint32_t>();
+  // Client asks for a description of a statement or a portal
+  // Statement descriptions (S) are returned as two separate messages: ParameterDescription and RowDescription
+  // Portal descriptions are just RowDescriptions
+  const auto description_target = _read_buffer.template get_value<char>();
+
+  Assert(description_target == 'P', "Only portal descriptions are currently supported.");
+
+  // The description itself will be sent out after execution of the prepared statement.
+  /* const auto statement_or_portal_name = */ _read_buffer.get_string(packet_length - LENGTH_FIELD_SIZE - sizeof(char));
+}
```

Comment:
The client asks for a description of a statement or portal (== the layout of the result table). At this point, the LQP has been instantiated and translated but hasn't been executed yet. Since I found no way to retrieve the layout of the result table from the LQP I decided to do it after its execution. To answer your questions:

>  Does it modify the state somehow ...

No.

> ... how does it have an effect?

The message has to be read at some point. We also have to check if the client requests a statement description. If there's a way to solve the problem stated above I can implement the logic here.

Reply:
> The description itself will be sent out after execution of the prepared statement.

Is this something that is independent of read_describe_packet? If yes, maybe add „in any case, ...“ and clarify that read_describe_packet has no effect.

If I am still mistaken, let‘s take this offline